pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -793,6 +793,27 @@ void updateInstanceStatesForNewConsumingSegment(Map<String, Map<String, String>>
       LOGGER.info("Updating segment: {} to ONLINE state", committingSegmentName);
     }
 
+    // There used to be a race condition in pinot (caused by heavy GC on the controller during segment commit)
```

Comment:
Is this caused by controller running into GC and leader changed, and 2 controllers trying to create the segment with the same seq num? If so, this change won't really work because the `instanceStatesMap` won't be up-to-date

Reply:
Last time this happened, it was caused by gc on the controller, and the same controller waking up to complete the operation it is scheduled to do. 
I could not dig up the PR since it happened (I think) when the source code was under com.linkedin domain, but it was fixed by @jackjlli  by adding a Stat check when we update metadata for the committing segment.

Reply:
I think it can only happen in the following orders:
- The current leader controller start committing a segment
- The current leader controller runs into GC and not responding, leadership changes to another controller
- The new leader controller committed the segment
- The previous leader controller wakes up and commit the segment again

If this is the order, the fix won't work because the `instanceStatesMap` is not up-to-date if it is already calculated before the controller runs into GC. The right fix might be checking the leadership again before actually committing the segment.

@jackjlli What was the fix?

Reply:
The fix is the usage of `Stat` class. This is the PR I made(https://github.com/apache/incubator-pinot/pull/2361). Now it's in `persistSegmentZKMetadata` method. 

If the former lead controller tried to persist its change, it'd fail since the stat version has already been changed.

Reply:
In any case, it picks up the latest Idealstate and checks on that znode. If a different controller added a segment, then even if the old controller gets back control, it will not be able to add the new segment